database/src/desktop/connection/persistent_connection.cc

```diff
@@ -551,28 +620,58 @@ void PersistentConnection::OnTokenFutureComplete(
   if (!lock.GetReference()) return;
 
   {
-    // Clear pending future
+    // Update your own status, and check if Auth is finished.
     MutexLock future_lock(connection->pending_token_future_mutex_);
-    connection->pending_token_future_ = Future<std::string>();
+    // If this future doesn't match the pending future, a different set is
+    // underway.
+    if (connection->pending_app_check_token_future_ != result_data) {
+      return;
+    }
+    connection->app_check_token_future_status_ = kCompletedTokenFuture;
+    if (connection->auth_token_future_status_ == kWaitingForTokenFuture) {
+      // Still waiting for Auth, so return and let the Auth
+      // callback finish the connection.
+      return;
+    }
   }
 
-  connection->scheduler_->Schedule(
-      new callback::CallbackValue2<ThisRef, Future<std::string>>(
-          connection->safe_this_, result_data,
-          [](ThisRef ref, Future<std::string> future) {
-            ThisRefLock lock(&ref);
-            if (lock.GetReference()) {
-              lock.GetReference()->HandleTokenFuture(future);
-            }
-          }));
+  connection->scheduler_->Schedule(new callback::CallbackValue1<ThisRef>(
+      connection->safe_this_, [](ThisRef ref) {
+        ThisRefLock lock(&ref);
+        if (lock.GetReference()) {
+          lock.GetReference()->HandleTokenFutures();
+        }
+      }));
 }
 
-void PersistentConnection::HandleTokenFuture(Future<std::string> future) {
-  if (future.error() == 0) {
+void PersistentConnection::HandleTokenFutures() {
+  FIREBASE_DEV_ASSERT(auth_token_future_status_ != kWaitingForTokenFuture);
+  FIREBASE_DEV_ASSERT(app_check_token_future_status_ != kWaitingForTokenFuture);
+
+  bool auth_error = auth_token_future_status_ == kCompletedTokenFuture &&
+                    pending_auth_token_future_.error() != 0;
+
+  if (auth_error) {
+    // Only care about Auth errors, because App Check errors are handled by the
```

Comment:
Could this be changed in the future?

Even though it is designed this way now, error handling App Check Token future seem to be safer

Reply:
It shouldn't.  The connection will fail to be made due to the lack of App Check token, so that part is guaranteed.  I agree it is a bit weird that it doesn't care about the error, but this is another bit of logic that is pulled from the Android/iOS implementations, so for consistency sake I feel like we should match it (the comment is even pulled from them :) )